feat: Make cert lifetime configurable (#306)
stackabletech/secret-operator: docs/modules/secret-operator/pages/secretclass.adoc

We generally aim to use as short-lived certificates as possible.
The problem is, that short lifetime periods imply frequent restart of services, which may be totally unnoticeable for some products,
annoying for others or devastation for products with a single point of failure and no recovery from outages (e.g. Trino coordinator).

We have spent some time thinking about this and did reach the following compromise:

1. To enforce security constraints, cluster administrators can set a maximum allowed certificate lifetime
   on the SecretClass issuing the certificates (defaults to 15d).
2. We default to a certificate lifetime of 24h.
3. Pods consuming the certificate can request a longer lifetime or shutdown expiration buffer via annotations
   on the xref:volume.adoc[Volume]. If they request a longer lifetime than the SecretClass allows,
   it will be shortened to the maximum allowed lifetime.
4. The Pods will be evicted 6h before the certificate expires, to ensure that no Pods are left running with expired secrets. Consumers can override this buffer using Volume annotations.
   This buffer must be long enough that the product is guaranteed to gracefully shut down.

> Most of our product operators will not add any annotation to the Volume, so the (short) default certificate lifetime applies.

Author: Thinking about it, that sentence was wrong.
Changed it to

> Most of our product operators will not set any specific certificate lifetime, so the default applies.

xref:scope.adoc[Scopes] are used to populate the claims (such as `subjectAlternateName`) of the provisioned certificates.

> ==== Certification lifetime

Reviewer: ```suggestion
==== Certificate lifetime
```

Reviewer: Most of this should eventually apply to all kinds of secrets, not just certificates.

Reviewer: But maybe we don't need to generalize the docs quite yet

> We generally aim to use as shot lived certificates as possible.

Reviewer: ```suggestion
We generally aim to use as short-lived certificates as possible.
```

The problem is, that short lifetime periods imply frequent restart of services, which may be totally unnoticeable for some products,
annoying for others or devastation for products with a single point of failure and no recovery from outages (e.g. Trino coordinator).

We have spent some time thinking about this and did reach the following compromise:

1. To enforce security constraints, cluster administrators can set a maximum allowed certificate lifetime

>    on the SecretClass issuing the certificates (we default to 15d otherwise).

Reviewer: ```suggestion
   on the SecretClass issuing the certificates (defaults to 15d).
```

2. We default to a certificate lifetime of 24h.
3. Pods consuming the certificate can request a longer lifetime or shutdown expiration buffer via annotations
   on the xref:volume.adoc[Volume]. If they request a longer lifetime than the SecretClass allows,
   it will be shortened to the maximum allowed lifetime.

> 4. As a default, 6h before the certificate expires the Pods will get evicted so no Pods are running with outdated
>    certificates. Consumers can override the buffer using annotations on the Volume as well.
>    This buffer needs to be big enough, so that the product is guaranteed to gracefully shut down.

Reviewer: ```suggestion
4. The Pods will be evicted 6h before the certificate expires, to ensure that no Pods are left running with expired secrets. Consumers can override this buffer using Volume annotations.
   This buffer must be long enough that the product is guaranteed to gracefully shut down.
```

          name: secret-provisioner-tls-ca
          namespace: default
        autoGenerate: true
      maxCertificateLifetime: 15d # optional
----

`autoTls`:: Declares that the `autoTls` backend is used.
`autoTls.ca`:: Configures the certificate authority used to issue `Pod` certificates.
`autoTls.ca.secret`:: Reference (`name` and `namespace`) to a K8s `Secret` object where the CA certificate and key is stored in the keys `ca.crt`
                      and `ca.key` respectively.
`autoTls.ca.autoGenerate`:: Whether the certificate authority should be provisioned if it can not be found.

> `autoTls.maxCertificateLifetime`:: Maximum lifetime the created certificates are allowed to have. In case consumers request a longer lifetime than allowed by this setting, the lifetime will be the minimum of both, so this setting takes precedence.

Reviewer: ```suggestion
`autoTls.maxCertificateLifetime`:: Maximum lifetime the created certificates are allowed to have. In case consumers request a longer lifetime than allowed by this setting, the lifetime will be the minimum of both.
```
